062-support-tiered-storage.md

```diff
@@ -0,0 +1,121 @@
+#  Support tiered storage natively in Strimzi
+
+## Current situation
+
+Kafka introduced tiered storage feature since 3.6. The feature is in early stage and it’s nice to support tiered storage configuration natively via Strimzi.
+
+## Motivation
+
+Supporting tiered storage via Strimzi allow Strimzi users to enable tiered storage natively with Strimzi config. By allowing to config custom configuration for RemoteStorageManager and RemoteLogMetadataManager, users are able to adopt tiered storage feature on Strimzi easily.
+
+It should also be noted that, it’s on the end-user to ensure that their tiered storage setup works with Strimzi; the operator itself is strictly responsible for pushing down the necessary config to the broker pods.
+
+## Proposal
+
+### RemoteStorage
+
+Add a RemoteStorage class under storage config (under https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaClusterSpec-schema-reference), which would support configuration for all tiered storage related configuration.
+
+The RemoteStorage class contains configuration for various aspects, including system config, RSM specific config, RLMM specific config.
+
+```
+kafka:
+  storage:
+    remote:
+      enabled: true
+      rlm:
+        # See below detail
+      rsm:
+        # See below detail
+      rlmm:
+        # See below detail
+```
+
+### RLM
+
+The RLM config supports `default` type only.  The config allows users to specify basic configuration for RemoteLogManager in Kafka:
+
+```
+      rlm:
+        threadPoolSize:
+        reader.threads:
+        reader.maxPendingTasks:
+```
+
+### RSM
+
+The RSM config only supports custom type. The config allows users to specify the RSM class information as well as other properties. 
+
+Because the `RemoteStorageManager` dependency is provided by Kafka plugin, the implementation could vary case by case. It’s important to allow user to pass in additional configuration values within the supplied prefix range, thus config field is provided to pass in and generate such properties.
+
+```
+      rsm:
+        type: custom
+        class.name:
+        class.path:  
+        impl.prefix: rsm.config.
+        config:
+          property1:
+          property2:
+          ...
+```
+
+### RLMM
```

Comment:
Why do you need the custom metadata manager? I think you should simply say that this supports only implementations based on the `TopicBasedRemoteLogMetadataManager` and that way hide all of this. Or do you really need support for a custom manager? Then you cannot bootstrap the security easily.

Reply:
For us we use the default implementation, but I know other companies like AWS are using their own custom implementation. The custom metadata config is definitely useful for future user adoption.

Another reason is, I need to allow customization on metadata manager's internal Kafka client config. In theory that can also be done via kafka.config, but again, I put it here for clarify and easier maintenance. 

Let me know if you are ok with keep it here or not, I am happy to adjust.

Reply:
> For us we use the default implementation, but I know other companies like AWS are using their own custom implementation. The custom metadata config is definitely useful for future user adoption.

I do not think AWS uses Strimzi. If they do and need this, they can comment on it. It problem I see is that we should not overcomplicate the design to match some virtual scenarios. Also, if you use custom metadata manager, Stirmzi will not know how to bootstrap it, so not sure what value does this option give you.

> Another reason is, I need to allow customization on metadata manager's internal Kafka client config. In theory that can also be done via kafka.config, but again, I put it here for clarify and easier maintenance.

Why do you need that? What do you need to customize and how? Maybe some intro to how the tiered storage works, what are the two different parts and what might be the use cases for different configurations would be useful as an intro to this proposal to bring everyone to the same understanding what and why is needed.

But from my perspective, we want to make this a stable API where the operator does some of the configuration work to guarantee that it will work in the long term. But that always requires some opinionated configuration and does not give you complete freedom. So there are tradeoffs you need to make. So we need to have clarity at what customizations are allowed and which are not.

Reply:
I revise the config now so these config are not in the tieredstorage section.

For your information, the RLMM implementation contains a Kafka client talking to local Kafka cluster. We need to customize some configuration of the Kafka client based on various need, and that is probably unique to our use case, e.g. on some SSL config. We can customize that via `spec.kafka.config` as well.